source/base/Text.ooc

```diff
@@ -0,0 +1,275 @@
+		this endsWith(This new(other))
+	}
+	endsWith: func ~char (c: Char) -> Bool {
+		(this length > 0) && (this _buffer[this length - 1] == c)
+	}
+	find: func ~char (c: Char, start := 0) -> Int {
+		result := -1
+		for (i in start .. this length)
+			if (this[i] == c) {
+				result = i
+				break
+			}
+		result
+	}
+	find: func ~text (text: This, start := 0) -> Int {
+		result := -1
+		for (i in start .. this length - text length + 1)
+			if (this slice(i, text length) == text) {
+				result = i
+				break
+			}
+		result
+	}
+	find: func ~string (string: String, start := 0) -> Int {
+		this find(This new(string), start)
+	}
+	take: func@ -> Bool {
+		this _changeOwnership(Owner Caller)
+	}
+	give: func -> This {
+		result := this copy()
+		result _changeOwnership(Owner Callee)
+		result
+	}
+	free: func@ -> Bool {
+		result: Bool
+		if (result = (this _owner != Owner Literal && this _owner != Owner Stack && this _buffer != null)) {
+			gc_free(this _buffer)
+			this _buffer = null
+			this _length = 0
+		}
+		result
+	}
+	free: func@ ~withCriteria (criteria: Owner) -> Bool {
+		this _owner == criteria && this free()
+	}
+	slice: func (start, distance: Int) -> This {
+		length := abs(distance)
+		if (start < 0)
+			start = this length + start
+		if (distance < 0)
+			start -= length
+		if (start < this length)
+			This new(this _buffer + start, Int minimum~two(length, this length - start), Owner Literal)
+		else
+			This new()
+	}
+	toString: func -> String {
+		String new(this _buffer, this length)
+	}
+	toInt: func -> Int {
+		this toLLong() as Int
+	}
+	toLong: func -> Long {
+		this toLLong() as Long
+	}
+	toLLong: func -> LLong {
+		this toLLong~inBase(this _detectNumericBase())
+	}
+	toInt: func ~inBase (base: Int) -> Int {
+		this toLLong~inBase(base) as Int
+	}
+	toLong: func ~inBase (base: Int) -> LLong {
+		this toLLong~inBase(base) as Long
+	}
+	toLLong: func ~inBase (base: Int) -> LLong {
+		result := 0 as LLong
+		if (this empty == false) {
+			if (this[0] == '-')
+				result = -1 * this slice(1, this length - 1) toULong~inBase(base)
+			else
+				result = this toULong~inBase(base) as LLong
+		}
+		result
+	}
+	toULong: func -> ULong {
+		this toULong~inBase(this _detectNumericBase())
+	}
+	toULong: func ~inBase (base: Int) -> ULong {
+		result := 0 as ULong
+		if (this empty == false) {
+			lastValidIndex := -1
+			start := 0
+			if (base == 16 && this startsWith(This new(c"0x", 2))) {
+				start += 2
+				lastValidIndex += 2
+			}
+			for (i in start .. this length) {
+				if (This _isNumeric(this[i], base))
+					++lastValidIndex
+				else
+					break
+			}
+			power := 1 as ULong
+			while (lastValidIndex >= 0) {
+				result += power * This _toInt(this[lastValidIndex], base)
+				--lastValidIndex
+				power *= base
+			}
+		}
+		result
+	}
+	toFloat: func -> Float {
+		this toLDouble() as Float
+	}
+	toDouble: func -> Double {
+		this toLDouble() as Double
+	}
+	toLDouble: func -> LDouble {
```

Comment:
Unless we change `LDouble` to `LongDouble`, I don't think `toLongDouble` makes a lot of sense.